### PR TITLE
Windows: Respect integrated GPU preference in Windows Settings

### LINKED ISSUE #81870: Godot Editor engages dedicated 3d card on laptop, against both windows and nvidia driver settings.
### Godot version

v4.1.1.stable.mono.official [bd6af8e0e]

### System information

Godot v4.1.1.stable.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2060 with Max-Q Design (NVIDIA; 31.0.15.3187) - Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz (16 Threads)

### Issue description

When developing on my laptop, I force engines to use the intel integrated video card to save battery when I am not plugged in. 

I do this either with Windows' Graphics Settings Performance Preferences (Godot executable set to "power saving"), or NVidia's own Control Panel (Manage 3D settings -> Program Settings -> Godot executable set to prefer Integrated Graphics.)

However, neither of these systems work. They work for most software and game engines (including Unity). 

I have found that if you manually change your project settings renderer to GL_Compatibility, the editor will no longer engage the dedicated graphics card, however, this is a project level setting and is not something a user should have to swap in order to choose a graphics card on a laptop. 

### Steps to reproduce

No sample code needed, on a laptop with both dedicated and integrated graphics, launch godot, create a new project, choose any renderer other than 'compatibility', note that dedicated 3d card is now engaged. 

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by alvinhochun**
Tested locally with Vulkan.

Fixes #81870
Related to #58430

### CODE REVIEW

**Comment by alvinhochun** [comment]
Using Sysinternals Process Monitor I can observe that `opengl32.dll` reads the same registry key as this PR. It also reads the value `DirectXUserGlobalSettings`, however setting it manually doesn't seem to affect the GPU selection when using `gl_compatibilty`. So I'd wager this is good enough at least for when running as non-packaged desktop app.

However, I neglected that Windows Settings also allows selecting the preference for a packaged app, and the current change doesn't handle this case. It is slightly more annoying to test but implementation should be trivial (using `GetCurrentPackageFamilyName`).

**Comment by alvinhochun** [comment]
> Windows Settings also allows selecting the preference for a packaged app

This is now implemented, and I've verified that it works by hijacking another packaged app.

File: platform/windows/os_windows.cpp [hunk 1]
```diff
@@ -622,6 +622,74 @@ Vector<String> OS_Windows::get_video_adapter_driver_info() const {
...
+			GetCurrentApplicationUserModelIdPtr GetCurrentApplicationUserModelId = (GetCurrentApplicationUserModelIdPtr)GetProcAddress(kernel32, "GetCurrentApplicationUserModelId");
+
+			if (GetCurrentApplicationUserModelId) {
+				UINT32 length = sizeof(value_name) / sizeof(value_name[0]);
+				LONG result = GetCurrentApplicationUserModelId(&length, value_name);
+				if (result == ERROR_SUCCESS) {
+					is_packaged = true;
+				} else if (result != APPMODEL_ERROR_NO_APPLICATION) {
+					return false;
+				}
```
Comment (bruvzg) [inline platform/windows/os_windows.cpp:651]: What's the point of this branch, the only other documented error is `ERROR_INSUFFICIENT_BUFFER` (which should not happen, ID is limited to 128 characters). In any case, why not to fallback to EXE name?

File: platform/windows/os_windows.cpp [hunk 2]
```diff
@@ -622,6 +622,74 @@ Vector<String> OS_Windows::get_video_adapter_driver_info() const {
...
+
+	LPCWSTR subkey = L"SOFTWARE\\Microsoft\\DirectX\\UserGpuPreferences";
+	HKEY hkey = nullptr;
+	LSTATUS result = RegOpenKeyExW(HKEY_CURRENT_USER, subkey, 0, KEY_READ, &hkey);
+	if (result != ERROR_SUCCESS) {
+		return false;
+	}
+
+	DWORD size = 0;
+	result = RegGetValueW(hkey, nullptr, value_name, RRF_RT_REG_SZ, nullptr, nullptr, &size);
```
Comment (bruvzg) [inline platform/windows/os_windows.cpp:668]: Reg. key name size is limited to 16K chars, I wounder what would happen if you try to set GPU preferences for app with longer path (will it truncate it or won't set it at all).

Probably not relevant in practice, Godot won't handle long paths without https://github.com/godotengine/godot/pull/91902, and `CreateProcess` can't handle long current directory paths, so it's unlikely will ever be over 260.

**Review by bruvzg — APPROVED** [review]

File: platform/windows/os_windows.cpp (hunk 1 above)
Comment (alvinhochun) [reply platform/windows/os_windows.cpp:651]: I guess I can just remove this branch. It's not like it's actually going to have any effect in practice.

File: platform/windows/os_windows.cpp (hunk 2 above)
Comment (alvinhochun) [reply platform/windows/os_windows.cpp:668]: Interesting point to consider. Though things are really whacky even when I make the Godot exe path just 256 chars -- it crashes when using the NVIDIA GPU *and* explorer.exe; even if I set it to use the Intel GPU (OpenGL selection is handled by the system) it still crashed explorer.exe (though it did run the project manager).

I think I am comfortable with leaving the code like this. It is obviously not a supported scenario (unless MS decides to raise the limit of the value name, which I doubt would ever happen) and `RegOpenKeyExW` should just return an error and the function returns false.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**